File: apple/VocabKit/Tests/VocabKitTests/StudyQueueTests.swift
import XCTest
@testable import VocabKit

final class StudyQueueTests: XCTestCase {
    let now = Date(timeIntervalSince1970: 1_700_000_000)

    private func word(_ lemma: String, lookupCount: Int) -> WordEntry {
        WordEntry(
            lemma: lemma,
            text: lemma,
            kind: .word,
            lookupCount: lookupCount,
            firstSeen: 0,
            lastSeen: 0,
            learnerDefinition: "def of \(lemma)",
            generatedBy: "test",
            schemaVersion: SCHEMA_VERSION
        )
    }

    func testDueCardsComeBeforeFreshAndExcludeNotDue() {
        let words = [
            word("alpha", lookupCount: 5),   // fresh
            word("bravo", lookupCount: 10),  // fresh, higher count
            word("charlie", lookupCount: 1), // due yesterday
            word("delta", lookupCount: 99),  // not due (tomorrow)
        ]
        let states: [String: SRSState] = [
            "charlie": SRSState(dueDate: SRSEngine.millis(now.addingTimeInterval(-86_400))),
            "delta": SRSState(dueDate: SRSEngine.millis(now.addingTimeInterval(86_400))),
        ]

        let queue = StudyQueue.build(words: words, states: states, now: now)
        let order = queue.map(\.entry.lemma)

        // charlie is due; delta is excluded; fresh words ordered by lookupCount desc.
        XCTAssertEqual(order, ["charlie", "bravo", "alpha"])
    }

    func testDueCardsSortedByDueDate() {
        let words = [word("a", lookupCount: 1), word("b", lookupCount: 1)]
        let states: [String: SRSState] = [
            "a": SRSState(dueDate: SRSEngine.millis(now.addingTimeInterval(-3_600))),
            "b": SRSState(dueDate: SRSEngine.millis(now.addingTimeInterval(-7_200))),
        ]
        let queue = StudyQueue.build(words: words, states: states, now: now)
        XCTAssertEqual(queue.map(\.entry.lemma), ["b", "a"]) // older due first
    }

    func testLimitTruncatesQueue() {
        let words = (0..<10).map { word("w\($0)", lookupCount: $0) }
        let queue = StudyQueue.build(words: words, states: [:], now: now, limit: 3)
        XCTAssertEqual(queue.count, 3)
        // Highest lookupCount first among fresh cards.
        XCTAssertEqual(queue.map(\.entry.lemma), ["w9", "w8", "w7"])
    }
}

File: windows/Vocb.Core.Tests/StudyQueueTests.cs
using Vocb.Core;
using Xunit;

namespace Vocb.Core.Tests;

public class StudyQueueTests
{
    private const long Now = 1_700_000_000_000L;

    private static WordEntry Word(string lemma, int lookupCount) => new()
    {
        Lemma = lemma,
        Text = lemma,
        Kind = WordKind.Word,
        LookupCount = lookupCount,
        LearnerDefinition = $"def of {lemma}",
        GeneratedBy = "test",
        SchemaVersion = 1
    };

    [Fact]
    public void DueCardsComeBeforeFresh_AndExcludeNotDue()
    {
        var words = new[]
        {
            Word("alpha", 5),    // fresh
            Word("bravo", 10),   // fresh, higher count
            Word("charlie", 1),  // due yesterday
            Word("delta", 99),   // not due (tomorrow)
        };
        var states = new Dictionary<string, SrsState>
        {
            ["charlie"] = new SrsState { DueDate = Now - SrsEngine.MillisPerDay },
            ["delta"] = new SrsState { DueDate = Now + SrsEngine.MillisPerDay },
        };

        var order = StudyQueue.Build(words, states, Now).Select(c => c.Entry.Lemma).ToArray();
        Assert.Equal(new[] { "charlie", "bravo", "alpha" }, order);
    }

    [Fact]
    public void DueCards_SortedByDueDate()
    {
        var words = new[] { Word("a", 1), Word("b", 1) };
        var states = new Dictionary<string, SrsState>
        {
            ["a"] = new SrsState { DueDate = Now - 3_600_000 },
            ["b"] = new SrsState { DueDate = Now - 7_200_000 },
        };
        var order = StudyQueue.Build(words, states, Now).Select(c => c.Entry.Lemma).ToArray();
        Assert.Equal(new[] { "b", "a" }, order);
    }

    [Fact]
    public void Limit_TruncatesQueue()
    {
        var words = Enumerable.Range(0, 10).Select(i => Word($"w{i}", i)).ToArray();
        var queue = StudyQueue.Build(words, new Dictionary<string, SrsState>(), Now, limit: 3);
        Assert.Equal(3, queue.Count);
        Assert.Equal(new[] { "w9", "w8", "w7" }, queue.Select(c => c.Entry.Lemma).ToArray());
    }
}
